### PR TITLE
Add support for ListenBrainz.org

### PR DESCRIPTION
**Pull request by treeshateorcs**

### CODE REVIEW

**Comment by treeshateorcs** [comment]
now this is a full fledged, working, not glitching alternative to https://github.com/mariusor/mpris-scrobbler/ in just a few hundreds of SLOC (that program is 8k SLOC)

**Comment by InputUsername** [comment]
Awesome, looks good! I will review this tomorrow when I have time, I'll look at listenbrainz-rust as well. Thanks for the contributions!

**Comment by treeshateorcs** [comment]
glad to be part of something good :) take your time

File: src/mainloop.rs [hunk 1]
```diff
@@ -146,13 +147,23 @@ pub fn run(config: &Config, scrobbler: &Scrobbler) {
...
                 if length > MIN_LENGTH && current_play_time > min_play_time {
-                    let scrobble = Scrobble::new(artist, title, album);
+                    let scrobble = Scrobble::new(artist.clone(), title.clone(), album.clone());
 
                     match scrobbler.scrobble(scrobble) {
                         Ok(_) => println!("Track scrobbled successfully"),
                         Err(err) => eprintln!("Failed to scrobble song: {}", err),
                     }
-
+                    if config.lb_token != "" {
```
Comment (InputUsername) [inline src/mainloop.rs]: `lb_token` seems optional, so maybe it's a good idea to wrap it in an `Option`. That way you can use matching: `if let Some(token) = config.lb_token`. This also means it is not required in the config file.

File: src/mainloop.rs [hunk 2]
```diff
@@ -146,13 +147,23 @@ pub fn run(config: &Config, scrobbler: &Scrobbler) {
...
+                        let listen = Listen {
+                            artist: &artist[..],
+                            track: &title[..],
+                            album: &album[..],
+                        };
+                        match listen.single(&config.lb_token[..]) {
+                            Ok(_) => println!("Track listened successfully"),
```
Comment (InputUsername) [inline src/mainloop.rs]: This wording (and also on the `Err` case) is a bit vague, maybe clarify that it's for ListenBrainz as opposed to Last.fm

File: rescrobbled.service [hunk 3]
```diff
@@ -0,0 +1,11 @@
+[Unit]
```
Comment (InputUsername) [inline rescrobbled.service]: Would it be possible to split this off to a separate PR? Since it's not really related to ListenBrainz support.

Still very useful though :smiley:

File: src/mainloop.rs [hunk 4]
```diff
@@ -169,12 +180,23 @@ pub fn run(config: &Config, scrobbler: &Scrobbler) {
...
                 Err(err) => eprintln!("Failed to update status: {}", err),
             }
+            if config.lb_token != "" {
+                let listen = Listen {
+                    artist: &artist[..],
+                    track: &title[..],
+                    album: &album[..],
+                };
+                match listen.playing_now(&config.lb_token[..]) {
+                    Ok(_) => println!("Status listened successfully"),
```
Comment (InputUsername) [inline src/mainloop.rs]: Same as above, this wording is a bit vague.

File: src/mainloop.rs (hunk 2 above)
Comment (treeshateorcs) [reply src/mainloop.rs]: "Track submitted to ListenBrainz successfuly" "Failed to submit track to ListenBrainz"
"Track submitted to Last.fm successfuly" "Failed to submit track to Last.fm"
are those okay?

Comment (treeshateorcs) [reply src/mainloop.rs]: "Status updated on Last.fm successfully" "Status updated on ListenBrainz successfully" 
"Failed to update status on Last.fm: {}" "Failed to update status on ListenBrainz: {}"

Comment (InputUsername) [reply src/mainloop.rs]: Perfect!

File: src/mainloop.rs [hunk 5]
```diff
@@ -150,18 +150,20 @@ pub fn run(config: &Config, scrobbler: &Scrobbler) {
                     let scrobble = Scrobble::new(artist.clone(), title.clone(), album.clone());
 
                     match scrobbler.scrobble(scrobble) {
-                        Ok(_) => println!("Track scrobbled successfully"),
-                        Err(err) => eprintln!("Failed to scrobble song: {}", err),
+                        Ok(_) => println!("Track submitted to Last.fm successfully"),
+                        Err(err) => eprintln!("Failed to submit track to Last.fm: {}", err),
                     }
-                    if config.lb_token != "" {
+                    if config.lb_token.clone().unwrap_or_default() != "" {
```
Comment (InputUsername) [inline src/mainloop.rs]: Using `if let Some(lb_token) = lb_token` is more idiomatic I think.

File: src/mainloop.rs [hunk 6]
```diff
@@ -183,18 +185,18 @@ pub fn run(config: &Config, scrobbler: &Scrobbler) {
             let scrobble = Scrobble::new(artist.clone(), title.clone(), album.clone());
 
             match scrobbler.now_playing(scrobble) {
-                Ok(_) => println!("Status updated successfully"),
-                Err(err) => eprintln!("Failed to update status: {}", err),
+                Ok(_) => println!("Status updated on Last.fm successfully"),
+                Err(err) => eprintln!("Failed to update status on Last.fm: {}", err),
             }
-            if config.lb_token != "" {
+            if config.lb_token.clone().unwrap_or_default() != "" {
```
Comment (InputUsername) [inline src/mainloop.rs]: Same as above.

File: src/mainloop.rs (hunk 5 above)
Comment (treeshateorcs) [reply src/mainloop.rs]: I don't understand :( 
i wrote
`            if let Some(config.lb_token) = config.lb_token {`
and it spits
` error| expected one of `)`, `,`, or `@`, found `.``

Comment (InputUsername) [reply src/mainloop.rs]: Sorry, I was unclear about this. `Some(config.lb_token)` should be `Some(ref token)`.

What this does is use pattern matching against `config.lb_token` (an option). If it's `Some`, it will set `token` to a reference to `config.lb_token`. Then you can simply pass `token` to `listen.playing_now` (because Rust can automatically dereference `&String` to `&str`).

More info about `if let`: https://doc.rust-lang.org/book/ch06-03-if-let.html

**Comment by treeshateorcs** [comment]
done
